### ISSUE TITLE
diverging_sub_expression false positive?

### ISSUE THREAD

**Issue by codingcampbell** [Manishearth/rust-clippy#1245]
Rust beginner here, so very possible I am writing this poorly, but it seems like this rule might be hitting a false positive with nested matches. Quick version info:

>
rustc 1.14.0-nightly (86affcdf6 2016-09-28)
cargo 0.13.0-nightly (19cfb67 2016-09-28)
clippy v0.0.91

And here's a snippet of the code (part of a larger function that returns `Result<Self, String>`):

```
let pixels = match reader.next_row() {
    Ok(row) => match row {
        Some(pixels) => pixels,
        None => return Err("Could not decode PNG: Blank image?".to_string()),
    },
    Err(err) => return Err(format!("Could not decode PNG: {}", err.to_string())),
};
```

(Some context: this is [next_row()](https://docs.rs/png/0.5.2/png/struct.Reader.html#method.next_row) from the png crate)

clippy yields this error:

```
warning: sub-expression diverges, #[warn(diverging_sub_expression)] on by default
 |
 | None => return Err("Could not decode PNG: Blank image?".to_string()),
 | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

I would really like to `return` out of the function here if there's a problem, which doesn't seem like poor form to me. I think `try!` does a very similar thing. Am I missing a better way to write this, or is it a false positive?

Thank you! cc @oli-obk

**Comment by oli-obk** [comment]
Hmm apparently I screwed up nested matches. I'll investigate. Thanks for the detailed report.

Side note: you can rewrite this as

```rust
match foo {
    Ok(Some(pixels)) => pixels,
    Ok(None) => return Err(...),
    Err(...) => return Err(...),
}
```
